### PR TITLE
backport: bitcoin#16528, #18027, #18782, #18787, #18805, #18888, #19502, #19077, #20125, #20153, #20198, #20262, #20266, #23608, #29510  - native descriptor wallets

### PR DESCRIPTION
**Pull request by knst**
## Issue being fixed or feature implemented
This PR is a batch of backports and related fixes to add a support of native descriptor wallets to Dash Core.

There're more related backports, but this PR is a minimal package of backports to get descriptor wallets working and unit/functional tests to succeed. To do: bitcoin#20226, bitcoin#21049, bitcoin#18788, bitcoin#20267, bitcoin#19230, bitcoin#19239, bitcoin#19441, bitcoin#19568, bitcoin#19979, bitcoin-core/gui#96, bitcoin#19136, bitcoin#21277, bitcoin#21063, bitcoin#21302, bitcoin#19651, bitcoin#20191, bitcoin#22446 and other.

Prior work: 
 - https://github.com/dashpay/dash/pull/5580
 - https://github.com/dashpay/dash/pull/5807

## What was done?
backports:
 - bitcoin/bitcoin#16528
 - bitcoin/bitcoin#18027
 - bitcoin/bitcoin#18805
 - bitcoin/bitcoin#18782
 - bitcoin/bitcoin#18787
 - bitcoin/bitcoin#20266
 - bitcoin/bitcoin#20153
 - bitcoin/bitcoin#18888
 - bitcoin/bitcoin#20198
 - bitcoin/bitcoin#20125
 - bitcoin/bitcoin#20262
 - bitcoin/bitcoin#23608
 - bitcoin/bitcoin#19077
 - bitcoin/bitcoin#19502
 - bitcoin/bitcoin#29510

and extra fixes and missing changes for bitcoin#20156, bitcoin#20202, bitcoin#20267, bitcoin#21634 + fix of auto-backup for sqlite wallets.

## How Has This Been Tested?
There're 2 new functional tests:  `wallet_importdescriptors.py` and `wallet_descriptor.py`
Beside that many functional tests run twice now: using legacy wallet and descriptor wallets: `wallet_hd.py`, `wallet_basic.py`, `wallet_labels.py`, `wallet_keypool_topup.py`, `wallet_avoidreuse.py`, `rpc_psbt.py`, `wallet_keypool_hd.py`, `rpc_createmultisig.py`, `wallet_encryption.py`.
With bitcoin#18788 expected to more tests run.

## Breaking Changes
N/A

## Checklist:
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added or updated relevant unit/integration/functional/e2e tests
- [x] I have made corresponding changes to the documentation
- [x] I have assigned this pull request to a milestone

### CODE REVIEW

**Review by UdjinM6 — CHANGES_REQUESTED** [review]
Looks good overall, few notes. Will test it a bit.

File: src/rpc/client.cpp
Comment (knst) [reply src/rpc/client.cpp:213]: > 16528: Note: this is a breaking change for createwallet rpc

should we swap no5 and no6 such as
```
    { "createwallet", 5, "load_on_startup"},
    { "createwallet", 6, "descriptors"},
```
to prevent breaking changes?

If there's any other breaking changes then no need workarounds like that.

Comment (knst) [reply src/rpc/client.cpp]: re-ordered: https://github.com/dashpay/dash/pull/5579/commits/c243ea9096dbb865f41d3dee18950185134439cd

**Comment by knst** [comment]
> @[knst](https://github.com/knst) knst [force-pushed](https://github.com/dashpay/dash/compare/c243ea9096dbb865f41d3dee18950185134439cd..e0fb0f4806aa11f2de5740faf7ab44508d77b63e) the bp-descriptors-1 branch from[ c243ea9 ](https://github.com/dashpay/dash/commit/c243ea9096dbb865f41d3dee18950185134439cd)to e0fb0f4


resolved conflict due to `a50f20d779 backport: partial Merge bitcoin-core/gui#96: Slight improve create wallet dialog`

**Review by UdjinM6 — APPROVED** [review]
light ACK

File: doc/release-notes-16528.md
Comment (kwvg) [inline doc/release-notes-16528.md:80]: Should the following Dash-specific RPCs also be disabled for descriptor wallets?

* `importelectrumwallet` (would require additional logic to transform into descriptors)
* `upgradetohd` (does current HD wallet functionality play well with descriptors as a hybrid wallet?)
* `dumphdinfo` (see above)

File: doc/release-notes-16528.md
Comment (kwvg) [inline doc/release-notes-16528.md:8]: nit: Dashify?

File: doc/release-notes-16528.md
Comment (kwvg) [inline doc/release-notes-16528.md:10]: nit: Dash switched to using major versions with v18, IIRC these changes will land as part of a 20.x release.

File: doc/release-notes-16528.md
Comment (kwvg) [inline doc/release-notes-16528.md:51]: nit: Should we drop the example if it's not applicable?

File: doc/release-notes-16528.md
Comment (kwvg) [inline doc/release-notes-16528.md:106]: nit: Consider dropping `wpkh` example (or use `pkh` instead)

File: src/wallet/wallet.cpp
Comment (kwvg) [inline src/wallet/wallet.cpp:326]: Is my understanding correct that descriptor wallets will be sans HD?

File: test/functional/feature_filelock.py
Comment (kwvg) [inline test/functional/feature_filelock.py:18]: nit: Consider squashing all bitcoin#20267-related TODOs into one commit, should make it easier to revert (same for TODOs hinging on other backports)

**Review by kwvg — CHANGES_REQUESTED** [review]

File: src/wallet/wallet.cpp
Comment (knst) [reply src/wallet/wallet.cpp:326]: Note: HD wallets have deterministic way to calculate private keys from a seed, but non-HD wallets have completely random keys which makes difficult to backup them or to use as hardware wallets.

**Descriptors wallets are descriptor hierarchical deterministic wallets (HD wallets).**

The difference between legacy wallets (current one) and descriptor wallets is how they keep private keys and scripts.

Legacy wallets creates a bunch of private keys (even HD wallets still keep 1000 of keys and an information how they are derived to be sure that same private key won't be re-used due to re-using same derivation). 

Descriptor wallets instead keep a ranges, such as (0, 1000) + derivation path to use.

There's one more thing is different which is important in terms of implementation but not useful at the moment:
descriptor wallets have that many different types of descriptor how many types of addresses wallet support. For Bitcoin descriptor wallets have 3 types of descriptors: `pkh()/44`, `sh(wpkh())/49`, `wpkh()/84'` for Dash is only `pkh()/44'`. So, legacy wallets keep only private keys and no information which type of address should be used, for descriptor wallets descriptor enforces to use one and only one type of address. that's not important at the moment, but if we would introduce any new type of address (such as Shnor/taproot addresses), our descriptor wallets will have extra type of descriptor then.

File: doc/release-notes-16528.md
Comment (knst) [reply doc/release-notes-16528.md:80]: > upgradetohd (does current HD wallet functionality play well with descriptors as a hybrid wallet?)

I think that's opposite impossible to have non-hd wallet for descriptor wallet,but `upgradetohd` lets also to change seed, so, I guess it should be enabled.

`dumphdinfo` also should be available as well.

`importelectrumwallet` is added to a list of exception

File: test/functional/feature_filelock.py
Comment (knst) [reply test/functional/feature_filelock.py:18]: grouped 4 commits for 20267

File: src/wallet/wallet.cpp
Comment (kwvg) [reply src/wallet/wallet.cpp:326]: Thanks for clarifying, great to know it's using BIP44! I was concerned that it might use the old derivation system by default (having a consistent one is especially important when that `migratewallet` RPC come over since it'd be odd if legacy wallets migrated have a BIP44 derivation but fresh descriptor wallets have a BIP32 derivation) but that's been put to rest.

One small change needed, the derivation path needs to be `44'/5'/0'/0`, not `44'/0'/0'/0` (the second number is the coin discriminator and Dash has been assigned the value `5`, see [SatoshiLabs registration](https://github.com/satoshilabs/slips/blame/f246c8b4861ecfb53f0741904bab4fc624a9fb4c/slip-0044.md#L36)). This probably got skipped over since all testnets have been assigned `1`.

**Descriptor wallets:**

HD descriptor wallets with Bitcoin's derivation path

**Legacy wallets:**

getwalletinfo output for legacy wallet

dumpwallet output for legacy wallet

---

The value can be fetched from `Params().ExtCoinType()`

Comment (knst) [reply src/wallet/wallet.cpp:326]: good catch @kwvg! Thanks for paying attention to derivation path. Fixed in the latest push

**Review by UdjinM6 — CHANGES_REQUESTED** [review]
pls see below + 28bb5f896101a1df90c65df7b2b12827e2eb96bc (and maybe 64889849e9, not sure if I've done it correctly)

**Comment by knst** [comment]
> pls see below + https://github.com/dashpay/dash/commit/28bb5f896101a1df90c65df7b2b12827e2eb96bc (and maybe https://github.com/dashpay/dash/commit/64889849e978c9ee85176ee00bf201a5c1ebd032, not sure if I've done it correctly)

I prepared proper fix for `internal keys` count, see https://github.com/dashpay/dash/pull/5579/commits/0325bf5a3695a98cd43ce3e5d436320d4da3dc7c


I [splitted ](https://github.com/dashpay/dash/commit/64889849e978c9ee85176ee00bf201a5c1ebd032) to 2 commits: one is dashification, one squashed to backport, other is separate commit

File: doc/release-notes-16528.md
Comment (knst) [reply doc/release-notes-16528.md]: updated to 20.1 for now. I will bump later to 20.2 or 21 depends how fast 20.1 will be released.

**Review by UdjinM6 — CHANGES_REQUESTED** [review]

**Review by UdjinM6 — DISMISSED** [review]
LGTM 👍 utACK

**Comment by UdjinM6** [comment]
pls consider adding https://github.com/bitcoin/bitcoin/pull/29510 here or in #5911

**Review by UdjinM6 — DISMISSED** [review]
utACK

**Review by kwvg — CHANGES_REQUESTED** [review]
The ability to generate receiving addresses in the GUI seems to be broken. `getnewaddress` seems unaffected and works just fine. (Tested on d6c97e639089be881d84eb5f7a2f1cbe8512da6c)


**`develop` (77019488):**

https://github.com/dashpay/dash/assets/63189531/f65b4554-bc83-4674-af26-493c417bc17f

**Latest changes (d6c97e63):**

https://github.com/dashpay/dash/assets/63189531/f8bbd900-ed60-490f-8e52-d71d5c66b7b9

**Comment by knst** [comment]
I confirm issue, due to different implementation of 
```
bool getNewDestination(const OutputType type, const std::string label, CTxDestination& dest) override
    {
        LOCK(m_wallet->cs_wallet);
        std::string error;
        return m_wallet->GetNewDestination(type, label, dest, error);
    }
```
vs
```
    bool getNewDestination(const std::string label, CTxDestination& dest) override         
    {
        auto spk_man = m_wallet->GetLegacyScriptPubKeyMan();                               
        if (!spk_man) {
            return false;                                                                  
        }   
        std::string error;                                                                 
        return m_wallet->GetNewDestination(label, dest, error);                            
    }
```
in Dash it works only with **legacy** script pub key man which is wrong for descriptor wallet case. I will prepare fix later after validation that there's no any other similar misusage around.

**Comment by knst** [comment]
> The ability to generate receiving addresses in the GUI seems to be broken.

please, check the latest version, seems fixed.

**Review by UdjinM6 — APPROVED** [review]
GUI works for me 👍 

light ACK

File: src/wallet/interfaces.cpp
Comment (UdjinM6) [inline src/wallet/interfaces.cpp:157]: `cs_wallet` is locked inside `GetNewDestination`, no need to lock it here
```suggestion
```

**Review by UdjinM6 — CHANGES_REQUESTED** [review]
actually... :)

File: src/wallet/interfaces.cpp
Comment (knst) [reply src/wallet/interfaces.cpp:157]: I keep it for unification with bitcoin's codebase. Even now [in the master](https://github.com/bitcoin/bitcoin/blob/master/src/wallet/interfaces.cpp#L156-L157) they still have this `cs_wallet` locked.
```
    util::Result<CTxDestination> getNewDestination(const OutputType type, const std::string& label) override
    {
        LOCK(m_wallet->cs_wallet);
        return m_wallet->GetNewDestination(type, label);
    }
```

Comment (UdjinM6) [reply src/wallet/interfaces.cpp:157]: oh, ok, nvm then

**Review by UdjinM6 — APPROVED** [review]
re-ACK

**Review by kwvg — CHANGES_REQUESTED** [review]
CoinJoin believes that it currently doesn't have enough keys to start a session and seems unable to do automatic backups* (Tested on 4dfe75722df64b6c81aef1ade000a790686ceb41). Legacy wallets seem unaffected.

_\* - I'm aware HD wallets don't need automatic backups_

**Latest changes (4dfe75722df64b6c81aef1ade000a790686ceb41):**

https://github.com/dashpay/dash/assets/63189531/f5e10409-3839-4dfb-8264-c849490ad17a

Debug Log:

```
2024-03-03T09:22:24Z Using SQLite Version 3.38.5
2024-03-03T09:22:24Z Using wallet /home/eclair/.dashcore/testnet3/wallets/Meow
2024-03-03T09:22:24Z init message: Loading wallet...
2024-03-03T09:22:24Z [Meow] Automatic wallet backups are not supported!
2024-03-03T09:22:24Z [Meow] External scriptPubKey Manager does not exist
2024-03-03T09:22:24Z [Meow] Internal scriptPubKey Manager does not exist
2024-03-03T09:22:24Z [Meow] nKeysLeftSinceAutoBackup: 0
2024-03-03T09:22:24Z [Meow] Wallet File Version = 10500
2024-03-03T09:22:24Z [Meow] Keys: 0 plaintext, 0 encrypted, 0 total; Watch scripts: 0; HD PubKeys: 0; Metadata: 0; Unknown wallet records: 0
2024-03-03T09:22:24Z [Meow] Automatic wallet backups are not supported!
2024-03-03T09:22:24Z [Meow] External scriptPubKey Manager does not exist
2024-03-03T09:22:24Z [Meow] Internal scriptPubKey Manager does not exist
2024-03-03T09:22:24Z [Meow] Wallet completed loading in             137ms
2024-03-03T09:22:24Z [Meow] External scriptPubKey Manager does not exist
2024-03-03T09:22:24Z [Meow] Internal scriptPubKey Manager does not exist
2024-03-03T09:22:24Z [Meow] External scriptPubKey Manager does not exist
2024-03-03T09:22:24Z [Meow] Internal scriptPubKey Manager does not exist
2024-03-03T09:22:24Z [Meow] setExternalKeyPool.size() = 0
2024-03-03T09:22:24Z [Meow] External scriptPubKey Manager does not exist
2024-03-03T09:22:24Z [Meow] Internal scriptPubKey Manager does not exist
2024-03-03T09:22:24Z [Meow] GetKeyPoolSize() = 0
2024-03-03T09:22:24Z [Meow] mapWallet.size() = 0
2024-03-03T09:22:24Z [Meow] m_address_book.size() = 0
2024-03-03T09:22:24Z [Meow] Set HD by default
2024-03-03T09:22:25Z [Meow] Encrypting Wallet with an nDeriveIterations of 337300
2024-03-03T09:22:25Z [Meow] Setting spkMan to active: id = 649af9d73b3b413de40d04aed4f5b7300a44b89f29f18cefe249c64704f31c25, type = 0, internal = 0
2024-03-03T09:22:25Z [Meow] Setting spkMan to active: id = 545966e11f0115d43ccc936a1a4cd06be8a852fe60d221b595fc190801f89941, type = 0, internal = 1
2024-03-03T09:22:25Z [Meow] Automatic wallet backups are not supported!
2024-03-03T09:22:25Z CoinJoin: autostart=0, multisession=0,sessions=4, rounds=4, amount=1000, denoms_goal=50, denoms_hardcap=300
[...]
2024-03-03T09:22:53Z [Meow] AddToWallet c709a3a1e5c4f00b3f717c6d53f5a1d1335e3bc476775dd24da3564b7954ad0b  new
[...]
2024-03-03T09:23:05Z [Meow] AddToWallet c709a3a1e5c4f00b3f717c6d53f5a1d1335e3bc476775dd24da3564b7954ad0b  update
2024-03-03T09:23:15Z [Meow] Automatic wallet backups are not supported!
[...]
2024-03-03T09:23:17Z [Meow] Automatic wallet backups are not supported!
2024-03-03T09:23:18Z [Meow] Automatic wallet backups are not supported!
2024-03-03T09:23:19Z [Meow] Automatic wallet backups are not supported!
2024-03-03T09:23:20Z [Meow] Automatic wallet backups are not supported!
2024-03-03T09:23:23Z [Meow] Automatic wallet backups are not supported!
2024-03-03T09:23:24Z [Meow] Automatic wallet backups are not supported!
2024-03-03T09:23:25Z [Meow] Automatic wallet backups are not supported!
2024-03-03T09:23:26Z [Meow] Automatic wallet backups are not supported!
2024-03-03T09:23:27Z [Meow] Automatic wallet backups are not supported!
2024-03-03T09:23:28Z [Meow] Automatic wallet backups are not supported!
2024-03-03T09:23:29Z [Meow] Automatic wallet backups are not supported!
```

**`develop` (5f6ae3ca14f5e68c77203fe946e90b0e51ad9e45):**

_Note: Footage sped up during restarting client, syncing and CoinJoin session_

https://github.com/dashpay/dash/assets/63189531/e2bdbf7c-826e-4660-81a0-c942f46428d4

`getwalletinfo` output (difficult to read in footage because sped up):

```json
{
  "walletname": "Mrrp",
  "walletversion": 120200,
  "format": "bdb",
  "balance": 5.00000000,
  "coinjoin_balance": 0.00000000,
  "unconfirmed_balance": 0.00000000,
  "immature_balance": 0.00000000,
  "txcount": 1,
  "timefirstkey": 1709458871,
  "keypoololdest": 1709458872,
  "keypoolsize": 999,
  "keypoolsize_hd_internal": 1000,
  "keys_left": 999,
  "unlocked_until": 0,
  "paytxfee": 0.00000000,
  "hdchainid": "27f39cc00dfe50cfba5eea7c1e7ac76e96171d784a2afe0853e758f18a29b99b",
  "hdaccountcount": 1,
  "hdaccounts": [
    {
      "hdaccountindex": 0,
      "hdexternalkeyindex": 1000,
      "hdinternalkeyindex": 1000
    }
  ],
  "avoid_reuse": false,
  "scanning": false,
  "private_keys_enabled": true
}
```

**Comment by knst** [comment]
> CoinJoin believes that it currently doesn't have enough keys to start a session and seems unable to do automatic backups*, which seems to upset CoinJoin (Tested on https://github.com/dashpay/dash/commit/4dfe75722df64b6c81aef1ade000a790686ceb41). Legacy wallets seem unaffected.

@kwvg ,
Let pasta review this PR and get merged soon :) Let's keep this bug like out-of-scope for this PR.
I am blocked by this PR and can't do anything else because this PR is already super huge.

Firstly, "descriptor wallets" are still experimental feature
Secondly, as I stated in PR description, this PR is a  **minimal package of backports to get descriptor wallets working and unit/functional tests to succeed**.

So, having this PR merged, I can create more PRs with more fixes, improvement and not blocked by one single-entry point which blocks everything (this PR).


I already have a bunch of backports with continuous implementation of Decriptor Wallets to be merged: https://github.com/dashpay/dash/pull/5911 but blocked by this PR.

Beside this issue the `protx` functionality doesn't work properly with Descriptor wallets and it requires massive refactoring for llmq/MN implementation, because MN's implementation wanting to have a raw private keys and sign messages (see `protx_register_common_wrapper` in rpc/evo.cpp and other usages of LegacyScriptPubKeyMan).


By the way, if you have any idea how to fix Coin Join for descriptor wallets - can you prepare a PR? Otherwise I am going to work with that after `protx` fixes & refactorings.

**Comment by kwvg** [comment]
> Let pasta review this PR and get merged soon :) Let's keep this bug like out-of-scope for this PR.

Sounds fine by me.

> By the way, if you have any idea how to fix Coin Join for descriptor wallets - can you prepare a PR?

I suspect it's a matter of letting Qt know this is an HD wallet (the lingering "Make sure to encrypt your wallet [...]" message leads me to believe that it's treating the descriptor wallet as a non-HD wallet, triggering autobackups when it shouldn't apply)

> Beside this issue the protx functionality doesn't work properly with Descriptor wallets and it requires massive refactoring for llmq/MN implementation, because MN's implementation wanting to have a raw private keys and sign messages (see protx_register_common_wrapper in rpc/evo.cpp and other usages of LegacyScriptPubKeyMan).

I'll update the tracker issue (https://github.com/dashpay/dash-issues/issues/59) to document it. If there's anything else known broken (or anything that is discovered to be broken), then I'd recommend mentioning it there and mentioning the PR where it's resolved.

It would be valuable insight into the kinds of code changes needed to restore behaviour.

**Review by kwvg — APPROVED** [review]
ACK

**Comment by knst** [comment]
There're multiple force pushes due to:
 - rebase on top of develop after re-merging master
 - resolve conflict with #5854
 - fix compilation error due to #5901 (bitcoin#23653)
 - dropped workaround to make non-breaking v20.1 https://github.com/dashpay/dash/pull/5579/commits/365feb639b049b01eee7acf6fd4ba45168a72d73
 - updated v20.1 to v21 in release notes
 - commit "fix: counting calculation of internal keys for Descriptor Wallets" simplified by moving in history - no need to fix wallet_keypool_hd.py if do this commit before backport bitcoin#16528

File: doc/release-notes-16528.md
Comment (UdjinM6) [inline doc/release-notes-16528.md:30]: >dropped workaround to make non-breaking v20.1 https://github.com/dashpay/dash/commit/365feb639b049b01eee7acf6fd4ba45168a72d73

Should probably mention that it's a breaking change then

**Review by PastaPastaPasta — APPROVED** [review]
utACK 6b71f274ae48a0bb390cc47deb2fa15b61d2fd00

**Review by UdjinM6 — APPROVED** [review]
utACK

**Review by kwvg — APPROVED** [review]
utACK 6b71f274ae48a0bb390cc47deb2fa15b61d2fd00

**Pull request merged**
